pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,327 @@
+	autoDetectedProvider operator.Provider
+	status               status.StatusManager
+	clusterDomain        string
+}
+
+// Reconcile reconciles IP pools in the cluster.
+//
+// - Query desired IP pools (from Installation)
+// - Query existing IP pools owned by this controller
+// - Reconcile the differences
+// - Populate Installation status with ALL IP pools in the cluster.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling IP pools")
+
+	// TODO: Needed?
+	// newActiveCM, err := r.checkActive(reqLogger)
+	// if err != nil {
+	// 	return reconcile.Result{}, err
+	// }
+
+	// Get the Installation object - this is the source of truth for IP pools managed by
+	// this controller.
+	instance := &operator.Installation{}
+	if err := r.client.Get(ctx, utils.DefaultInstanceKey, instance); err != nil {
+		if apierrors.IsNotFound(err) {
+			reqLogger.Info("Installation config not found")
+			r.status.OnCRNotFound()
+			return reconcile.Result{}, nil
+		}
+		reqLogger.Error(err, "An error occurred when querying the Installation resource")
+		return reconcile.Result{}, err
+	}
+
+	// Mark CR found so we can report converter problems via tigerastatus
+	r.status.OnCRFound()
+
+	// Get the APIServer. If healthy, we'll use it for managing pools.
+	// Otherwise, we'll use the CRD API for bootstrapping the cluster until the API server is available.
+	apiserver, _, err := utils.GetAPIServer(ctx, r.client)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return reconcile.Result{}, err
+	}
+	apiAvailable := apiserver != nil && apiserver.Status.State == v1.TigeraStatusReady
+
+	// SetMetaData in the TigeraStatus such as observedGenerations.
+	defer r.status.SetMetaData(&instance.ObjectMeta)
+
+	// Get all IP pools in the cluster.
+	currentPools := &crdv1.IPPoolList{}
+	err = r.client.List(ctx, currentPools)
+	if err != nil && !errors.IsNotFound(err) {
+		r.status.SetDegraded(operator.ResourceReadError, "error querying IP pools", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// Create a lookup map of pools owned by this controller for easy access.
+	// This controller will ignore any IP pools that it itself did not create.
+	ourPools := map[string]crdv1.IPPool{}
+	for _, p := range currentPools.Items {
+		// TODO: Check if owned by this installation resource / controller.
+		ourPools[p.Spec.CIDR] = p
+	}
+
+	// For each pool that is desired, but doesn't exist, create it.
+	toCreate := []client.Object{}
+	for _, p := range instance.Spec.CalicoNetwork.IPPools {
+		if pool, ok := ourPools[p.CIDR]; !ok || !reflect.DeepEqual(pool, p) {
+			// Create a new pool, or update an existing one.
+			res := p.ToCRD()
+			if apiAvailable {
+				toCreate = append(toCreate, crdToV3(res))
+			} else {
+				toCreate = append(toCreate, res)
```

Comment:
Is the reason we need to bypass the api server if it's not running because the api server can't be brought up initially without some IPPool configured for pod networking?

After the initial networking is setup, if the apiserver is down the user wouldn't be able to create or delete IPPools currently as it requires the apiserver, or is this incorrect?

Reply:
> Is the reason we need to bypass the api server if it's not running because the api server can't be brought up initially without some IPPool configured for pod networking?

Yeah, the Calico apiserver generally runs using pod networking and so in order to access the v3 API you need to have an IP pool created first. This is a bit of a chicken/egg situation. 

> After the initial networking is setup, if the apiserver is down the user wouldn't be able to create or delete IPPools currently as it requires the apiserver, or is this incorrect?

In the current PR, creation of IP pools will still succeed even if the apiserver is down. Deletion will never occur unless it can use the v3 API, but if the apiserver is down the pool will just be disabled instead (and eventually deleted when the apiserver is running again).